src/v/coding-style.md

```diff
@@ -73,130 +21,70 @@ absolute path with `""` like this:
+### Ordering
 
 ```c++
-void a_function() {
-    if (some condition) {
-        stmt;
-    } else {
-        stmt;
-    }
-}
+class type {
+public:
+protected:
+private:
+};
 ```
 
-An exception is namespaces -- the body is _not_ indented, to prevent files that are almost 100% whitespace left margin.
+### Duplicate sections
 
-When making a change, if you need to insert an indentation level, you can temporarily break the rules by inserting a half-indent, so that the patch is easily reviewable:
+It's okay to use multiple instances of access specifiers to create logical groupings where needed.
 
 ```c++
-void a_function() {
-  while (something) {   // new line - half indent
-    if (some condition) {
-        stmt;
-    } else {
-        stmt;
-    }
-  }                      // new line
-}
-```
-
-A follow-up patch can restore the indents without any functional changes.
-
-## Function parameters
-
-Avoid output parameters; use return values instead.  In/out parameters are tricky, but in some cases they are relatively standard, such as serialization/deserialization.
-
-If a function accepts a lambda or an `std::function`, make it the last argument, so that it can be easily provided inline:
-
-```c++ 
-template <typename Func>
-void function_accepting_a_lambda(int a, int b, Func func);
-
-int f() {
-    return function_accepting_a_lambda(2, 3, [] (int x, int y) {
-        return x + y;
-    });
-}
-```
-
-## Complex return types
+class type {
+public:
+   // boiler plate
 
-If a function returns a complicated return type, put its return type on a separate line, otherwise it becomes hard to see where the return type ends and where the function name begins:
+public:
+  // logical grouping of APIs
 
-```c++
-template <typename T1, T2>
-template <typename T3, T4>
-std::vector<typename a_struct<T1, T2>::some_nested_class<T3, T4>>  // I'm the return type
-a_struct<T1, T2>::a_function(T3 a, T4 b) {                         // And I'm the function name
-    // ...
-}
+...
+};
 ```
 
-## Whitespace around operators
+## Name Lookup
 
-Whitespace around operators should match their precedence: high precedence = no spaces, low precedency = add spaces:
+### Hidden friends
 
-```c++
-     return *a + *b;  // good
-     return * a+* b;  // bad
-```
-
-Parentheses can be used to make the expression clearer, but spaces should still
-be used for readability.
+https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1601r0.pdf
 
-`if`, `while`, `return` (and `template`) are not function calls, so they get a space after the keyword.
+With hidden friends the function lookup is found by ADL only; i.e., it cannot be found by unqualified lookup.
 
-## Long lines
-
-If a line becomes excessively long (>80 characters?), or is just complicated,
-break it into two or more lines.  The second (and succeeding lines) are
-_continuation lines_, and have a half indent:
+E.g. implement operators as friends, ideally inline:
```

Comment:
Why is `inline` ideal? I've been using non-inline to prevent from having to #include any fmtlib stuff in headers.

Reply:
I simply copied it from the github discussion. This goes back to your earlier point about not knowing what was "decided". Should I just remove even more from this guide so we can debate things more formally in separate PRs?

Reply:
Personally outside of `ideally` inline, I'm a huge +1 to all this here (although maybe in the naming section we need something about Google Test fixtures..). But let's maybe find the minimal amount of things here and merge it to update then we can figure out a process for updates going forward.

Reply:
Sounds like it's to do with "hidden friends", the general recommendation there is inline, but as far as I can tell, it only matters that the declaration is inline with the class, the definition can go anywhere. I.e., don't declare it at block scope and then declare it a friend in the class declaration, only declare it inline with the class.

The goal is to depend on ADL, which reduces the set of names during lookup (reduces compile time), and to prevent implicit conversions.

Reply:
sounds like we should remove inline or specify its declaration should be inline with the class declaration and the definition can go anywhere?

I am onboard with this